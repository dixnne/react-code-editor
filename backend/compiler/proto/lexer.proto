syntax = "proto3";

package compiler;

service Lexer {
  rpc Analyze(AnalyzeRequest) returns (TokenList);
}

message AnalyzeRequest {
  string input = 1;
}

message Token {
  string token_type = 1;
  string lexeme = 2;
  uint32 line = 3;
  uint32 column = 4;
}

message TokenList {
  repeated Token tokens = 1;
}