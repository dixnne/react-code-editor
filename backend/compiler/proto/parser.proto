syntax = "proto3";

package compiler;

import "lexer.proto";

service Parser {
  rpc Parse (ParseRequest) returns (ParseResponse);
  rpc ParseSource (ParseSourceRequest) returns (ParseResponse);
}

message ParseRequest {
  repeated Token tokens = 1;
}

message ParseSourceRequest {
  string source = 1;
}

message ParseResponse {
  ASTNode ast = 1;
  repeated ParserError errors = 2;
}

message ASTNode {
  string node_type = 1;
  string value = 2;
  repeated ASTNode children = 3;
  uint32 start_line = 4;
  uint32 start_column = 5;
  uint32 end_line = 6;
  uint32 end_column = 7;
}

message ParserError {
  string error_type = 1;
  string message = 2;
  uint32 line = 3;
  uint32 column = 4;
}