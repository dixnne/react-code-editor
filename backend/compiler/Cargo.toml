[package]
name = "compiler"
version = "0.1.0"
edition = "2024"


# Esta sección define tu biblioteca.
[lib]
name = "compiler"
path = "src/lib.rs"

# Esta sección define tu ejecutable.
[[bin]]
name = "compiler"
path = "src/main.rs"

[dependencies]
prost = "0.13.5"
tonic = { version = "0.12.0"}
tokio = { version = "1", features = ["macros", "rt-multi-thread"] }
tonic-reflection = "0.12.0"

[build-dependencies]
tonic-build = "0.10"

[dev-dependencies]
insta = { version = "1.34.0", features = ["yaml", "glob"] }