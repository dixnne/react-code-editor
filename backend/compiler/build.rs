fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Genera el código de Rust para los archivos .proto
    // y lo coloca en el directorio de salida estándar de Cargo (`OUT_DIR`).
    tonic_build::configure()
        .build_server(true)
        // La siguiente línea es opcional pero útil para herramientas de reflexión como grpcurl
        .file_descriptor_set_path(
            std::env::var("OUT_DIR").unwrap() + "/compiler_descriptor.bin",
        )
        .compile(
            &["proto/lexer.proto", "proto/parser.proto"], // Lista de archivos a compilar
            &["proto"], // Directorio donde buscar imports
        )?;

    Ok(())
}
