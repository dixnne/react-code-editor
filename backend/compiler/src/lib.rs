// Contenido de src/lib.rs (Corregido)

pub mod ast;
pub mod lexer;
pub mod parser;
pub mod token;