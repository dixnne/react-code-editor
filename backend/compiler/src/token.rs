use core::fmt;

// Se añade `Copy` para optimizar, ya que los enums son baratos de copiar.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenType {
    CommentSingle,    // Comentario de una sola línea (ej. // comentario)
    CommentMultiLine, // Comentario de múltiples líneas (ej. /* comentario */)
    Keyword,          // Palabra reservada del lenguaje (ej. let, if, while)
    Identifier,       // Nombre dado por el usuario a variables, funciones, etc.
    Integer,          // Número entero (ej. 10, 42)
    Float,            // Número de punto flotante (ej. 3.14, 0.5)
    String,           // Cadena de caracteres (ej. "hola", 'mundo')
    Boolean,          // Valor booleano (true, false)
    
    // --- Operadores ---
    Plus,             // Operador de suma (+)
    Minus,            // Operador de resta (-)
    Asterisk,         // Operador de multiplicación (*)
    Slash,            // Operador de división (/)
    Equal,            // Operador de asignación (=)
    Greater,          // Operador mayor que (>)
    Less,             // Operador menor que (<)
    Exclamation,      // Signo de exclamación (!)
    Ampersand,        // Ampersand (&) - Generalmente para operaciones a nivel de bit
    Bar,              // Barra vertical (|) - Generalmente para operaciones a nivel de bit
    DoubleEqual,      // Operador de igualdad (==)
    GreaterEqual,     // Operador mayor o igual que (>=)
    LessEqual,        // Operador menor o igual que (<=)
    NotEqual,         // Operador de desigualdad (!= o <>)
    DoubleAmpersand,  // Operador lógico AND (&&)
    DoubleBar,        // Operador lógico OR (||)
    Increment,        // Operador de incremento (++)
    Decrement,        // Operador de decremento (--)
    
    // --- Operadores Especiales (Nombres Corregidos) ---
    Splat,            // @*
    Spread,           // ...+
    Pipe,             // |>
    Swap,             // <=>  <-- AÑADIDO

    // --- Delimitadores y Puntuación ---
    ArrowRight,       // Flecha (->)
    LeftParen,        // Paréntesis izquierdo (()
    RightParen,       // Paréntesis derecho ())
    LeftBrace,        // Llave izquierda ({)
    RightBrace,       // Llave derecha (})
    LeftBracket,      // Corchete izquierdo ([)
    RightBracket,     // Corchete derecho (])
    Comma,            // Coma (,)
    Semicolon,        // Punto y coma (;)
    Colon,            // Dos puntos (:)
    Dot,              // Punto (.)
    
    // --- Tokens Misceláneos ---
    Whitespace,       // Espacio en blanco, tabulación, etc.
    NewLine,          // Salto de línea
    EndOfFile,        // Marcador de fin de archivo/entrada
    Unknown,          // Token desconocido/inválido
}

impl TokenType {
    /// Convierte un string a un TokenType.
    /// Necesario para la comunicación con gRPC.
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "Keyword" => Some(TokenType::Keyword),
            "Identifier" => Some(TokenType::Identifier),
            "Integer" => Some(TokenType::Integer),
            "Float" => Some(TokenType::Float),
            "String" => Some(TokenType::String),
            "Boolean" => Some(TokenType::Boolean),
            "Plus" => Some(TokenType::Plus),
            "Minus" => Some(TokenType::Minus),
            "Asterisk" => Some(TokenType::Asterisk),
            "Slash" => Some(TokenType::Slash),
            "Equal" => Some(TokenType::Equal),
            "DoubleEqual" => Some(TokenType::DoubleEqual),
            "NotEqual" => Some(TokenType::NotEqual),
            "Less" => Some(TokenType::Less),
            "LessEqual" => Some(TokenType::LessEqual),
            "Greater" => Some(TokenType::Greater),
            "GreaterEqual" => Some(TokenType::GreaterEqual),
            "DoubleAmpersand" => Some(TokenType::DoubleAmpersand),
            "DoubleBar" => Some(TokenType::DoubleBar),
            "Exclamation" => Some(TokenType::Exclamation),
            "Splat" => Some(TokenType::Splat),
            "Spread" => Some(TokenType::Spread),
            "Pipe" => Some(TokenType::Pipe),
            "Swap" => Some(TokenType::Swap),
            "Increment" => Some(TokenType::Increment),
            "Decrement" => Some(TokenType::Decrement),
            "LeftParen" => Some(TokenType::LeftParen),
            "RightParen" => Some(TokenType::RightParen),
            "LeftBrace" => Some(TokenType::LeftBrace),
            "RightBrace" => Some(TokenType::RightBrace),
            "LeftBracket" => Some(TokenType::LeftBracket),
            "RightBracket" => Some(TokenType::RightBracket),
            "Comma" => Some(TokenType::Comma),
            "Semicolon" => Some(TokenType::Semicolon),
            "Colon" => Some(TokenType::Colon),
            "Dot" => Some(TokenType::Dot),
            "ArrowRight" => Some(TokenType::ArrowRight),
            _ => None,
        }
    }
}

impl fmt::Display for TokenType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct LexerToken {
    pub token_type: TokenType,
    pub lexeme: String,
    pub line: usize,
    pub column: usize,
}

impl LexerToken {
    pub fn new(token_type: TokenType, lexeme: String, line: usize, column: usize) -> Self {
        Self {
            token_type,
            lexeme,
            line,
            column,
        }
    }
}
