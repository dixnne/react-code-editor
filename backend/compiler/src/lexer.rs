use std::str::Chars;
use std::iter::Peekable;
use crate::token::TokenType;
use std::collections::HashMap;
use crate::token::LexerToken;

pub struct LexicalAnalyzer<'a> {
    input: Peekable<Chars<'a>>,
    line: usize,
    column: usize,
    keywords: HashMap<String, TokenType>,
}

impl<'a> LexicalAnalyzer<'a> {
    pub fn new(source: &'a str) -> Self {
        let mut keywords = HashMap::new();
        keywords.insert("let".to_string(), TokenType::Keyword);
        keywords.insert("const".to_string(), TokenType::Keyword);
        keywords.insert("fn".to_string(), TokenType::Keyword);
        keywords.insert("if".to_string(), TokenType::Keyword);
        keywords.insert("else".to_string(), TokenType::Keyword);
        keywords.insert("while".to_string(), TokenType::Keyword);
        keywords.insert("struct".to_string(), TokenType::Keyword);
        keywords.insert("return".to_string(), TokenType::Keyword);
        keywords.insert("for".to_string(), TokenType::Keyword);
        keywords.insert("in".to_string(), TokenType::Keyword);
        keywords.insert("true".to_string(), TokenType::Boolean);
        keywords.insert("false".to_string(), TokenType::Boolean);

        Self {
            input: source.chars().peekable(),
            line: 1,
            column: 1,
            keywords,
        }
    }

    fn advance(&mut self) -> Option<char> {
        if let Some(ch) = self.input.next() {
            if ch == '\n' {
                self.line += 1;
                self.column = 1;
            } else {
                self.column += 1;
            }
            Some(ch)
        } else {
            None
        }
    }

    fn peek(&mut self) -> Option<&char> {
        self.input.peek()
    }

    fn is_at_end(&mut self) -> bool {
        self.peek().is_none()
    }

    fn match_next(&mut self, expected: char) -> bool {
        if self.peek() == Some(&expected) {
            self.advance();
            true
        } else {
            false
        }
    }

    pub fn scan_tokens(&mut self) -> Vec<LexerToken> {
        let mut tokens = Vec::new();
        while !self.is_at_end() {
            tokens.push(self.scan_token());
        }
        tokens.push(LexerToken::new(TokenType::EndOfFile, "".to_string(), self.line, self.column));
        tokens
    }

    fn scan_token(&mut self) -> LexerToken {
        let start_line = self.line;
        let start_column = self.column;

        let ch = match self.advance() {
            Some(c) => c,
            None => return LexerToken::new(TokenType::EndOfFile, "".to_string(), start_line, start_column),
        };

        match ch {
            ' ' | '\t' | '\r' => {
                let mut lexeme = String::from(ch);
                while let Some(' ' | '\t' | '\r') = self.peek() {
                    lexeme.push(self.advance().unwrap());
                }
                LexerToken::new(TokenType::Whitespace, lexeme, start_line, start_column)
            }
            '\n' => LexerToken::new(TokenType::NewLine, "\n".to_string(), start_line, start_column),
            '/' => {
                if self.match_next('/') {
                    let mut comment = String::new();
                    while let Some(c) = self.peek() {
                        if *c == '\n' { break; }
                        comment.push(self.advance().unwrap());
                    }
                    LexerToken::new(TokenType::CommentSingle, comment, start_line, start_column)
                } else if self.match_next('*') {
                    let mut comment = String::new();
                    while !self.is_at_end() {
                        if self.match_next('*') && self.match_next('/') {
                            return LexerToken::new(TokenType::CommentMultiLine, comment, start_line, start_column);
                        }
                        comment.push(self.advance().unwrap());
                    }
                    LexerToken::new(TokenType::Unknown, comment, start_line, start_column) // Unterminated comment
                } else {
                    LexerToken::new(TokenType::Slash, "/".to_string(), start_line, start_column)
                }
            }
            // --- Delimitadores y Operadores ---
            '(' => LexerToken::new(TokenType::LeftParen, "(".to_string(), start_line, start_column),
            ')' => LexerToken::new(TokenType::RightParen, ")".to_string(), start_line, start_column),
            '{' => LexerToken::new(TokenType::LeftBrace, "{".to_string(), start_line, start_column),
            '}' => LexerToken::new(TokenType::RightBrace, "}".to_string(), start_line, start_column),
            '[' => LexerToken::new(TokenType::LeftBracket, "[".to_string(), start_line, start_column),
            ']' => LexerToken::new(TokenType::RightBracket, "]".to_string(), start_line, start_column),
            ',' => LexerToken::new(TokenType::Comma, ",".to_string(), start_line, start_column),
            ';' => LexerToken::new(TokenType::Semicolon, ";".to_string(), start_line, start_column),
            ':' => LexerToken::new(TokenType::Colon, ":".to_string(), start_line, start_column),
            '.' => {
                if self.peek() == Some(&'.') {
                    self.advance();
                    if self.peek() == Some(&'.') {
                        self.advance();
                        if self.peek() == Some(&'+') {
                            self.advance();
                            return LexerToken::new(TokenType::Spread, "...+".to_string(), start_line, start_column);
                        }
                    }
                }
                LexerToken::new(TokenType::Dot, ".".to_string(), start_line, start_column)
            },
            '+' => if self.match_next('+') { LexerToken::new(TokenType::Increment, "++".to_string(), start_line, start_column) } else { LexerToken::new(TokenType::Plus, "+".to_string(), start_line, start_column) },
            '-' => if self.match_next('>') { LexerToken::new(TokenType::ArrowRight, "->".to_string(), start_line, start_column) } else if self.match_next('-') { LexerToken::new(TokenType::Decrement, "--".to_string(), start_line, start_column) } else { LexerToken::new(TokenType::Minus, "-".to_string(), start_line, start_column) },
            '*' => LexerToken::new(TokenType::Asterisk, "*".to_string(), start_line, start_column),
            '=' => if self.match_next('=') { LexerToken::new(TokenType::DoubleEqual, "==".to_string(), start_line, start_column) } else { LexerToken::new(TokenType::Equal, "=".to_string(), start_line, start_column) },
            '>' => if self.match_next('=') { LexerToken::new(TokenType::GreaterEqual, ">=".to_string(), start_line, start_column) } else { LexerToken::new(TokenType::Greater, ">".to_string(), start_line, start_column) },
            '<' => if self.match_next('=') {
                if self.match_next('>') { LexerToken::new(TokenType::Swap, "<=>".to_string(), start_line, start_column) }
                else { LexerToken::new(TokenType::LessEqual, "<=".to_string(), start_line, start_column) }
            } else if self.match_next('>') { LexerToken::new(TokenType::NotEqual, "<>".to_string(), start_line, start_column) } else { LexerToken::new(TokenType::Less, "<".to_string(), start_line, start_column) },
            '!' => if self.match_next('=') { LexerToken::new(TokenType::NotEqual, "!=".to_string(), start_line, start_column) } else { LexerToken::new(TokenType::Exclamation, "!".to_string(), start_line, start_column) },
            '&' => if self.match_next('&') { LexerToken::new(TokenType::DoubleAmpersand, "&&".to_string(), start_line, start_column) } else { LexerToken::new(TokenType::Ampersand, "&".to_string(), start_line, start_column) },
            '|' => if self.match_next('>') { LexerToken::new(TokenType::Pipe, "|>".to_string(), start_line, start_column) } else if self.match_next('|') { LexerToken::new(TokenType::DoubleBar, "||".to_string(), start_line, start_column) } else { LexerToken::new(TokenType::Bar, "|".to_string(), start_line, start_column) },
            '@' => if self.match_next('*') { LexerToken::new(TokenType::Splat, "@*".to_string(), start_line, start_column) } else { LexerToken::new(TokenType::Unknown, "@".to_string(), start_line, start_column) },
            // --- Literales ---
            '\'' | '"' => {
                let quote_char = ch;
                let mut content = String::new();
                while let Some(&next) = self.peek() {
                    if next == quote_char { break; }
                    if next == '\n' { return LexerToken::new(TokenType::Unknown, format!("{}{}", quote_char, content), start_line, start_column); } // Unterminated string
                    content.push(self.advance().unwrap());
                }
                self.advance(); // Consume closing quote
                LexerToken::new(TokenType::String, content, start_line, start_column)
            }
            c if c.is_alphabetic() || c == '_' => {
                let mut identifier = String::from(c);
                while let Some(&next) = self.peek() {
                    if next.is_alphanumeric() || next == '_' {
                        identifier.push(self.advance().unwrap());
                    } else { break; }
                }
                if let Some(token_type) = self.keywords.get(&identifier).cloned() {
                    LexerToken::new(token_type, identifier, start_line, start_column)
                } else {
                    LexerToken::new(TokenType::Identifier, identifier, start_line, start_column)
                }
            }
            c if c.is_digit(10) => {
                let mut number_str = String::from(c);
                while let Some(&next) = self.peek() {
                    if next.is_digit(10) {
                        number_str.push(self.advance().unwrap());
                    } else { break; }
                }
                if self.peek() == Some(&'.') {
                     if self.input.clone().nth(1).map_or(false, |c| c.is_digit(10)) {
                        number_str.push(self.advance().unwrap()); // Consume '.'
                        while let Some(&next) = self.peek() {
                           if next.is_digit(10) {
                               number_str.push(self.advance().unwrap());
                           } else { break; }
                        }
                        return LexerToken::new(TokenType::Float, number_str, start_line, start_column);
                     }
                }
                LexerToken::new(TokenType::Integer, number_str, start_line, start_column)
            }
            _ => LexerToken::new(TokenType::Unknown, ch.to_string(), start_line, start_column),
        }
    }
}
