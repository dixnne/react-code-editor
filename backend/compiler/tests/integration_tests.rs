// Usa el nombre del crate "compiler" para importar la lógica de la biblioteca
use compiler::lexer::LexicalAnalyzer;
use compiler::parser::parse_tokens;
use compiler::token::{LexerToken, TokenType};
use std::fs;
use std::path::Path;

fn run_snapshot_test(file_path: &str) {
    let path = Path::new(file_path);
    let source = fs::read_to_string(path)
        .unwrap_or_else(|_| panic!("No se pudo leer el archivo de prueba: {}", file_path));

    let mut lexer = LexicalAnalyzer::new(&source);
    let tokens = lexer.scan_tokens();

    let filtered_tokens: Vec<LexerToken> = tokens
        .into_iter()
        .filter(|t| !matches!(t.token_type, TokenType::Whitespace | TokenType::NewLine | TokenType::CommentSingle | TokenType::CommentMultiLine | TokenType::Unknown))
        .collect();
    
    let result = parse_tokens(&filtered_tokens);

    let ast_string = format!("{:#?}", result.ast);
    
    // El nombre del snapshot se deriva del nombre del archivo
    insta::assert_snapshot!(insta::internals::AutoName, ast_string, file_path);
}

// Creamos una función de prueba explícita para cada archivo de caso.
// Esto es más robusto que usar la macro `glob!`.
#[test]
fn test_simple_let() {
    run_snapshot_test("tests/cases/01_simple_let.dreamc");
}

#[test]
fn test_functions() {
    run_snapshot_test("tests/cases/02_functions.dreamc");
}

#[test]
fn test_syntax_error() {
    run_snapshot_test("tests/cases/03_syntax_error.dreamc");
}

// Puedes añadir más pruebas aquí para otros archivos...
// #[test]
// fn test_functions() {
//     run_snapshot_test("tests/cases/02_functions.dreamc");
// }
