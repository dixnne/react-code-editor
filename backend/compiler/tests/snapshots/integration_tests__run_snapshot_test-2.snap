---
source: tests/integration_tests.rs
expression: tests/cases/01_simple_let.dreamc
---
Program {
    declarations: [
        Variable(
            VariableDeclaration {
                identifier: Identifier {
                    name: "a",
                    line: 2,
                    column: 5,
                },
                var_type: Some(
                    Int,
                ),
                value: Literal(
                    Int(
                        10,
                    ),
                ),
            },
        ),
        Variable(
            VariableDeclaration {
                identifier: Identifier {
                    name: "b",
                    line: 3,
                    column: 5,
                },
                var_type: None,
                value: Binary {
                    left: Identifier(
                        Identifier {
                            name: "a",
                            line: 3,
                            column: 9,
                        },
                    ),
                    op: Plus,
                    right: Literal(
                        Int(
                            5,
                        ),
                    ),
                },
            },
        ),
    ],
}
